pkg/gatewayserver/gatewayserver.go

```diff
@@ -654,38 +654,51 @@ func (gs *GatewayServer) handleUpstream(conn connectionEntry) {
 	}
 }
 
-// UpdateConnectionStats updates the stats for a single gateway connection.
-func (gs *GatewayServer) UpdateConnectionStats(ctx context.Context, conn *io.Connection) error {
-	return gs.statsRegistry.Set(ctx, conn.Gateway().GatewayIdentifiers, conn.Stats())
+var allTraffic = io.Traffic{Up: true, Down: true, Status: true}
+
+// UpdateAllConnectionStats updates the stats for a single gateway connection.
+func (gs *GatewayServer) UpdateAllConnectionStats(ctx context.Context, conn *io.Connection) error {
+	return gs.statsRegistry.Set(ctx, conn.Gateway().GatewayIdentifiers, conn.Stats(), allTraffic)
 }
 
 func (gs *GatewayServer) updateConnStats(conn connectionEntry) {
 	ctx := conn.Context()
 	logger := log.FromContext(ctx)
 
+	if gs.statsRegistry == nil {
+		return
+	}
+
 	defer func() {
 		logger.Debug("Delete connection stats")
-		err := gs.statsRegistry.Set(ctx, conn.Gateway().GatewayIdentifiers, nil)
+		err := gs.statsRegistry.Set(ctx, conn.Gateway().GatewayIdentifiers, nil, io.Traffic{})
 		if err != nil {
 			logger.WithError(err).Error("Failed to delete connection stats")
 		}
 	}()
 	for {
+		var traffic io.Traffic
 		select {
 		case <-ctx.Done():
 			return
-		case <-conn.StatsChanged():
-			err := gs.UpdateConnectionStats(ctx, conn.Connection)
-			if err != nil {
-				logger.WithError(err).Error("Failed to update connection stats")
-			}
+		case <-conn.UpChanged():
+			traffic.Up = true
+		case <-conn.DownChanged():
+			traffic.Down = true
+		case <-conn.StatusChanged():
+			traffic.Status = true
+		}
 
-			timeout := time.After(gs.updateConnectionStatsDebounceTime)
-			select {
-			case <-ctx.Done():
-				return
-			case <-timeout:
-			}
+		err := gs.statsRegistry.Set(ctx, conn.Gateway().GatewayIdentifiers, conn.Stats(), conn.GetTraffic(traffic))
```

Comment:
You know from the local vars above which fields are set, right? You read those channels already, why read them again in `GetTraffic()`?

Reply:
As discussed, this was to drain updates of all message types reliably, removed since we use one channel only.